app/cluster/state/etcd.go

```diff
@@ -0,0 +1,289 @@
+				return fmt.Errorf("marshal ack value: %w", err)
+			}
+			_, err = manager.Client.Put(context.Background(), req.AckKey, ackValue)
+			if err != nil {
+				return fmt.Errorf("put value to ack key %q: %w", req.AckKey, err)
+			}
+			return err
+		})
+}
+
+func errChModeRequest(err error) <-chan servermode.ChangeEvent {
+	ch := make(chan servermode.ChangeEvent, 1)
+	ch <- servermode.ChangeEventError(err)
+	close(ch)
+	return ch
+}
+
+func (manager *ETCDManager) ServerMode(ctx context.Context) <-chan servermode.ChangeEvent {
+	if err := manager.init(); err != nil {
+		return errChModeRequest(err)
+	}
+
+	modeRequestKey := fmt.Sprintf(modeRequestKeyPattern, manager.Config.Namespace, manager.Config.ServerIndex)
+
+	resultChan := make(chan servermode.ChangeEvent, 1)
+	resp, err := manager.Client.Get(ctx, modeRequestKey)
+	if err != nil {
+		return errChModeRequest(err)
+	}
+
+	if len(resp.Kvs) == 0 {
+		return errChModeRequest(errors.New("no workspace found"))
+	}
+
+	resultChan <- manager.unmarshalMode(resp.Kvs[0].Value)
+
+	etcdWatchChan := manager.Client.Watch(ctx, modeRequestKey, clientv3.WithRev(resp.Header.Revision+1))
+	go func() {
+		for watchResp := range etcdWatchChan {
+			if watchResp.Err() != nil {
+				resultChan <- servermode.ChangeEventError(watchResp.Err())
+				continue
+			}
+
+			for _, event := range watchResp.Events {
+				switch event.Type {
+				case mvccpb.PUT:
+					resultChan <- manager.unmarshalMode(event.Kv.Value)
+				default:
+					resultChan <- servermode.ChangeEventError(fmt.Errorf("unknown event type %q", event.Type))
+				}
+			}
+		}
+		close(resultChan)
+	}()
+
+	return resultChan
+}
+
+func errChWorkspacesRequest(err error) <-chan workspace.ChangeEvent {
+	ch := make(chan workspace.ChangeEvent, 1)
+	ch <- workspace.ChangeEventError(err)
+	close(ch)
+	return ch
+}
+
+func (manager *ETCDManager) unmarshalWorkspace(raw []byte) workspace.ChangeEvent {
+	var req workspacesRequestsValue
+	err := json.Unmarshal(raw, &req)
+	if err != nil {
+		return workspace.ChangeEventError(err)
+	}
+
+	return workspace.NewWorkspacesRequest(
+		strings.Split(req.Workspaces, ","),
+		func() error {
+			ackValue, err := json.MarshalToString(workspacesAckValue{
+				Status: "RELOADED",
+			})
+			if err != nil {
+				return fmt.Errorf("marshal ack value: %w", err)
+			}
+			_, err = manager.Client.Put(context.Background(), req.AckKey, ackValue)
+			return err
+		})
+}
+
+func (manager *ETCDManager) WorkspaceIDs(ctx context.Context) <-chan workspace.ChangeEvent {
+	if err := manager.init(); err != nil {
+		return errChWorkspacesRequest(err)
+	}
+
+	modeRequestKey := fmt.Sprintf(workspacesRequestsKeyPattern, manager.Config.Namespace, manager.Config.ServerIndex)
+
+	resultChan := make(chan workspace.ChangeEvent, 1)
+	resp, err := manager.Client.Get(ctx, modeRequestKey)
+	if err != nil {
+		return errChWorkspacesRequest(err)
+	}
+
+	if len(resp.Kvs) == 0 {
+		return errChWorkspacesRequest(errors.New("no workspace found"))
```

Comment:
This might be a plausible case when we are using manual mode, We would be first scaling up the server with empty workspaces and then would move a workspace on to that empty server

Reply:
Should we be returning an error ? I feel we can just ignore and move ahead

Reply:
What if we wait until the key is available ? This is the latest change I made. Should we do the same for workspaceIDs ?

Reply:
Yeah workspaces and mode behave in a similar fashion so it should be same for both I guess

Reply:
will return an empty workspace